### ISSUE TITLE
SVG best practice for epub3

### ISSUE THREAD

**Issue by cortopy** [jgm/pandoc#2766]
I'm following "Epub 3 Best Practices" by Matt Garrish and Markus Gylling to find the best way to embed svg images in an epub3 file.

I'm trying to provide a fallback to PNG with Pandoc, but I'm unable to do so. The following is recommended as the best way to include svg pics in an epub:

```XML
<object data="image01.svg" type="image/svg+xml" width="640" height="480">
<img src="image01.png" alt="..." />
</object>
```

The reason being that you can show the svg for ereaders and browsers that support it and, at the same time, provide a fallback for those that don't.

However, the markdown link format can only create an `<img>`. If there isn't a method to achieve this, it would be great to tell Pandoc in the markdown that a given svg file has a png/jpeg fallback associated to it.

**Comment by mb21** [comment]
Actually, you can include [raw HTML](http://pandoc.org/README.html#raw-html):

```
<object data="image01.svg" type="image/svg+xml" width="640" height="480">
  ![](image01.png)
</object>
```

**Comment by cortopy** [comment]
But Pandoc doesn't interpret what's inside HTML items, so the svg would never be included in the epub and referenced in the manifest. The data attribute (and by extension the object element) would be meaningless I think

**Comment by mb21** [comment]
> But Pandoc doesn't interpret what's inside HTML items

yes it does, give it a try...

**Comment by mb21** [comment]
Hm... what about a hack like:

```
<object data="image01.svg" type="image/svg+xml" width="640" height="480">
  ![](image01.png)

  ![](image01.svg){width=0 height=0}
</object>
```

**Comment by jeffmcneill** [comment]
Commenting here to follow the answer...

**Comment by cortopy** [comment]
@mb21 I've just tested your solution but it's not working. Pandoc will rename all the attached images, whereas your data attribute value will remain "image01.svg", an asset which will not exist in the epub.
Not only that the output now has both object and image tags, so there's content duplication too

**Comment by mb21** [comment]
Maybe the HTML writer should output `<object data="foo.svg"><img src="foo.png"></object>` when encountering an image with an SVG suffix. Probably should check whether foo.png exists on the file system as well (before it gets put into the ePub)...

**Comment by ickc** [comment]
I can see why someone want it to be done like this, but should pandoc support this natively? The problems are:

1. should it give options for having svg alone, or both svg and a png fall back?
2. And if png fall back is offered (supposedly you provide an svg), should pandoc generates a png on the fly?

And then some might ask for a feature, in a certain sense similar to this, say responsive images. Then almost the same 2 questions can be asked again (should you give the highest resolution image and let pandoc generates the lower one on the fly?).

Because of the complication involved (e.g. feature like this won't please everyone so some would want to have options, then that options will further complicates everything), and this is more about html/epub design rather than providing a simple, valid output (svg is perfectly valid epub3, so it seems we are in the polyfill kind of fix), I don't think this is something that should be done in pandoc.

Then the question is how it is done outside pandoc: you write HTML in the source, and then use any softwares that can edit the zip (unpack and repack requires a particular MIME) and to inject the images there. The whole process can be written in a makefile so that the whole process can be automated when modified in the future.

**Comment by mb21** [comment]
Yes, you can always fix the ePUB after generation, but it would be nice if pandoc would just generate ePUBs that work on a majority of eBook readers and browsers. (SVGs are very common in HTML documents nowadays.)

So my proposal was to generate the `<object data="foo.svg"><img src="foo.png"></object>` _only_ if both foo.svg and foo.png are present and can be included in the ePUB by pandoc. I think it's a reasonable compromise in terms of complexity/convenience.

**Comment by ickc** [comment]
My problem with that is it isn't a fix, because it is not invalid.

And then this "best practice" is not a practice without drawback. Unlike for example delivering retina images through responsive images, the multiple versions of the images are on the server and the best one is served. The only extra "bits" they get is the HTML source is a bit longer, which hardly is a problem. But in this case, everyone receiving the epub is suffering: the majority of people receiving unnecessary and big pngs, the minority of people getting the svg for nothing. If one really cannot be sure if the readers on the clients' side can render SVG or not, ditching SVG might actually helps everyone (at least smaller than SVG+PNG). Or one can simply render 2 versions, one for SVG, one for PNG and let the end-users choose (I suppose if you have a lack of control on the readers' capability, probably you're giving it away for free. But even if it's under a paywall, it doesn't prevent you to deliver 2 versions.). Yes, none of what I suggested is perfect, but the "best" up there isn't perfect either, and then I would say it is subjective to say which one is the best.

And checking the existence of a png "sidecar" won't solve the whole problem, because people happen to have this sidecar will think this behavior is unexpected. And then people wanting this feature would actually want the png be auto-generated. If one need to write a script to pre-generate it anyway, this feature is not "complete" in the sense pandoc handling everything, then may be it's better for pandoc to not handle anything.

Just my 2 cents though.

**Comment by mb21** [comment]
I'd be interested to see which readers (and what usage percentage) support SVGs...

**Comment by ickc** [comment]
http://epubtest.org/testsuite/epub3/

**Comment by mb21** [comment]
thanks, any idea about the marketshare of those readers? pity that the Kindle's seem to be stuck at 26% SVG support...

**Comment by ickc** [comment]
IMO this issue itself is a "small" issue, that implement or not is not a big deal. So I am more concern about the implication: targetting a spec is "easy" enough, but targetting a changing world that has so much politics and most vendors like to play "let's wait and see" game is a very tough one.

And there's isn't no solution with the current pandoc: one can use pure PNG if they want to target more audiences (or like blindly target audiences without knowing who they are). Likewise if there's any epub3 feature that pandoc use and some reader don't read, use epub2 output instead.

So my point is to take away this burden from pandoc. If there's truly a need on this, rather than spending time to code it in pandoc (which also post extra constraint, like complexity, library dep cannot be too large, etc.), and probably has a half-solution (i.e. the png is not generated), may be we should implement this as a filter that can freely use any libraries/languages to serve the niche.

P.S.: I personally would really like to see pandoc more "modular" through the filter system, and so making filters more of a standard rather than only used by some geeks. (this is the idea behind the "ecosystem" thing. So that filters is not to avoid but part of a normal life around manipulating pandoc documents.)

Regarding the marketshare: I have no idea. But intuition (e.g. from news around Amazon and Apple's ebook dominance) suggests that the Amazon's Kindle book store and the Apple's iBook store dominate the market. And because of Android, Google Play Book store should come in a 3rd. In my mind things like Nook, etc. comes in a long distance from these 3. But I could be wrong.

And regarding the Kindle: does Kindle natively support epub? At least in the past, if one is to publish ebook on Kindle store, one have to produce the document in kindle's proprietory format, and usually people just convert it from, say, epub. From this alone (and if it is still true presently), Kindle's support is irrelevent to pandoc, because people have to take an extra build process to get to kindle format (and so one can tailor this separate build process to avoid the SVG problem).

**Comment by jgm** [comment]
+++ Mauro Bieg [Jan 22 17 04:45 ]:
>   So my proposal was to generate the <object data="foo.svg"><img
>   src="foo.png"></object> only if both foo.svg and foo.png are present
>   and can be included in the ePUB by pandoc. I think it's a reasonable
>   compromise in terms of complexity/convenience.

This makes sense to me.

**Comment by mb21** [comment]
I had a look at this but currently the HTML writer (which the ePUB writer calls) doesn't do IO, so no way to see whether the png exists... should I make a PR based on the typeclass branch or wait till that is merged into master?

**Comment by jgm** [comment]
Probably best to wait until it's in master.

+++ Mauro Bieg [Jan 28 17 07:16 ]:

**Comment by mb21** [comment]
see pull https://github.com/jgm/pandoc/pull/3396
